### ISSUE TITLE
[Request] module_variable_optional_attrs: Optional nested maps/objects?

### ISSUE THREAD

**Issue by skyzyx** [hashicorp/terraform#27613]
### Current Terraform Version

```
0.14.6
```

### Use-cases

I have a complex module. I would like to allow my end-users to be able to group like-variables together — specifically as a nested object/map. I'm looking at the `module_variable_optional_attrs` experiment and hoping it can help me simplify code.

### Attempted Solutions

I've tried defining a variable like this…

```hcl
variable "integrations" {
  description = "<p>Third-party integrations to configure.</p>"
  default     = {}

  type = optional(object({
    pagerduty = optional(object({
      prod    = optional(string)
      nonprod = optional(string)
    }))
  }))
}
```

…and using it like this…

```hcl
integrations = {
  "pagerduty" = {
    "prod" = "abcd1234"
  }
}
```

I received the error:

> Keyword "optional" is valid only as a modifier for object type attributes.

OK. So I removed the outer `optional(…)`s.

```hcl
variable "integrations" {
  description = "<p>Third-party integrations to configure.</p>"
  default     = {}

  type = object({
    pagerduty = object({
      prod    = optional(string)
      nonprod = optional(string)
    })
  })
}
```

I received the error:

> This default value is not compatible with the variable's type constraint: attribute "pagerduty" is required.

However, it appears that I can use it if I avoid the nesting.

```hcl
variable "pagerduty_integrations" {
  description = "<p>Third-party integrations to configure.</p>"
  default     = {}

  type = object({
    prod    = optional(string)
    nonprod = optional(string)
  })
}
```

I can use it like this:

```hcl
pagerduty_integrations = {
  "prod" = "abcd1234"
}
```

**Comment by apparentlymart** [comment]
Hi @skyzyx,

I think the original error was caused by you putting `optional` around the outermost type constraint. The syntax for making an entire variable optional is different because it predates there being this idea of optional object attributes: you can make it optional by giving it a `default`, which you did:

```hcl
variable "integrations" {
  description = "Third-party integrations to configure."
  type = object({
    pagerduty = optional(object({
      prod    = optional(string)
      nonprod = optional(string)
    }))
  })
  default = {}
}
```

I believe the above will work and get the result you were hoping for. Can you give it a try and let me know if there's still something missing? Thanks!

---

Side-note: the above is an interesting example of the behavior of optional object attributes: without `optional` that `default` value you specified would've been invalid, because it doesn't include a value for the `pagerduty` argument, but with `optional` it should now work and actually cause the default to _effectively_ be the following, due to type conversion:

```hcl
{
  pagerduty = null
}
```

**Comment by jalaziz** [comment]
I believe this is related to #28344.

**Comment by korinne** [comment]
Hi all, we recently released the [Terraform v1.3 alpha](https://releases.hashicorp.com/terraform/1.3.0-alpha20220608), which includes the ability to mark object type attributes as optional and set default values. 

I'm following up with issues that have provided feedback on the previous experiment, and wanted to invite you all to try the alpha and provide any feedback on this updated design. You can learn more/add comments[ here](https://discuss.hashicorp.com/t/request-for-feedback-optional-object-type-attributes-with-defaults-in-v1-3-alpha/40550). Thank you so much in advance, and hope you like the feature!

**Comment by doug-fish** [comment]
I'm having trouble taking advantage of this new capability for my use case. I make use of maps that include an arbitrary object name in the middle. Here's an example:

```
ases = {
  web = {
    size = "I1"
    networksize = 28
  },
  app2 = {
    size = "I1"
    networksize = 27
  }
}
```

The elements web, and app2 are arbitrary names that become object names. This pattern was useful for letting us do things like delete the web ase without any modification to the app2 ase occurring.

Currently I'm sometimes writing validation like this, but it's non-trivial.
```
variable "ases" {
  description = "a map of ase names => ase descriptor objects"
  validation {
    condition = (
      length(keys(var.ases)) > 0 &&
      alltrue([for ase in values(var.ases) :
        length(setsubtract(keys(ase), ["size", "networksize", "extra_nsg_rules"])) == 0 &&
        length(setsubtract(["size", "networksize""], keys(ase))) == 0
      ])
    )
    error_message = "Variable ases must be a map of the form { ASENAMES* = { size, networksize, ruleoffset, asp_size, asp_capacity, extra_nsg_rules(optional), remote_peer_backends(optional)."
  }
}
```

Is there any way to use the optional declarations to eliminate/simplify my validation?

**Comment by apparentlymart** [comment]
Hi all! I think we missed closing this one out as part of stabilizing optional attribute syntax with optional defaults in the main branch in preparation for the forthcoming v1.3 release, so I'm going to do that now.

However, before I do I'll quickly address @doug-fish's question...

I'm not sure I've fully understood your goal from what you described but inferring a little from the `error_message` specified in your validation rule in your second code snippet I think the corresponding type constraint under the newly-merged syntax (which will be available in v1.3.0, once released) would be like this:

```hcl
variable "ases" {
  type = map(
    object({
      size                 = string
      networksize          = number
      ruleoffset           = number
      asp_size             = string
      asp_capacity         = number
      extra_nsg_rules      = optional(set(string))
      remote_peer_backends = optional(set(string))
    })
  )
}
```

I had to guess a bit what would be good type constraints for each of the attributes here so I probably didn't get that exactly how you need it, but the important thing here is the `optional(...)` modifier around the type constraints for `extra_nsg_rules` and `remote_peer_backends`. With that extra modifier, Terraform will accept an input object that lacks either or both of those attributes and will quietly convert it to the correct type by setting them to `null`. You can then write the logic elsewhere in your module to specify what should happen when those are null.

You can also optionally specify a second argument to the `optional` modifier to provide a default value other than `null`. For example, it might be reasonable for the default value of `extra_nsg_rules` to be an empty set, which you can achieve by specifying an empty tuple as the default value and letting Terraform automatically convert it to match the `set(string)` type constraint:

```hcl
      extra_nsg_rules      = optional(set(string), [])
```

I'm going to close this issue now because the original request it represented has been met, but if you have any further questions about what I've described above feel free to start a topic about it in [the community forum](https://discuss.hashicorp.com/c/terraform-core/27). Thanks!
